src/main/java/org/truffleruby/core/numeric/IntegerNodes.java

```diff
@@ -495,15 +496,18 @@ public abstract static class IDivNode extends CoreMethodArrayArgumentsNode {
         @Specialization
         Object idiv(Object a, Object b,
                 @Cached InlinedConditionProfile zeroProfile,
-                @Cached FloatToIntegerNode floatToIntegerNode) {
+                @Cached FloatToIntegerNode floatToIntegerNode,
+                @Cached DispatchNode floorNode) {
             Object quotient = divNode.executeDiv(a, b);
             if (quotient instanceof Double) {
                 if (zeroProfile.profile(this, (double) b == 0.0)) {
                     throw new RaiseException(getContext(), coreExceptions().zeroDivisionError(this));
                 }
                 return floatToIntegerNode.execute(this, Math.floor((double) quotient));
-            } else {
+            } else if (RubyGuards.isImplicitLong(quotient)) {
```

Comment:
Might quotient be greater than max long? In this case `RubyGuards.isRubyInteger` should be used

Reply:
Good catch. I'll add a new spec.

Reply:
Never mind. The issue here isn't one of correctness. The fallback case handles a `RubyBignum` just fine; it's just not as efficient.